### PR TITLE
Beta feature: Multimodal topic modelling

### PR DESCRIPTION
**Pull request by x-tabdeveloping**
I'm currently implementing multimodal topic modelling in the library. This includes extending the encoder interface to be able to use multimodal embedding models, adding interpretation utilites and adding models.

### Tasks:

 1. [x] Implement multimodal encoder interface:
     - [x] sentence-transformers compatibility 
     - [x] MIEB compatibility
     - [x] Data and model validation
     - [ ] Handle missing images (***This turned out to be harder than I anticipated, probably left for future releases***)
2. [ ] Add multimodal models:
     - [x] KeyNMF
     - [x] $S^3$ 
     - [x] ZeroShotTM (***Slightly different from original, should perhaps fix for paper***)
     - [x] CombinedTM
     - [x] BERTopic
     - [x] Top2Vec
     - [x] GMM
     - [ ] Contrast ([paper](https://aclanthology.org/2024.lrec-main.1064/)) (***I will rather add this to `topic-benchmark` for the paper, but I don't like it enough to implement in Turftopic***)
3. [ ] Add interpretation utilites:
     - [x] Display topic images
     - [ ] Add images to compasses (***Future release***)
     - [ ] Add images to dynamic plots (***Future release***)
4. [x] Add docs
5. [x] Add Tests
6. [ ] Add docstrings

### Preview:

I have trained $S^3$ on an IKEA dataset:

```python
import mteb
import numpy as np
from datasets import load_dataset
from sklearn.feature_extraction.text import CountVectorizer

from turftopic import SemanticSignalSeparation

ds = load_dataset("nbadrinath/ikea_dataset_5.0", split="train[:1000]")
images = ds["image"]
texts = ds["desc"]

encoder = mteb.get_model("kakaobrain/align-base")

print("Embedding documents")
model = SemanticSignalSeparation(
    5,
    encoder=encoder,
    random_state=42,
    vectorizer=CountVectorizer(min_df=1, stop_words="english"),
)

doc_topic_matrix = model.fit_transform_multimodal(
    texts, images=images
)

fig = model.plot_topics_with_images()
fig.show()
```

image

### CODE REVIEW

**Comment by KennethEnevoldsen** [comment]
> (I will rather add this to topic-benchmark for the paper, but I don't like it enough to implement in Turftopic)

Perfectly fine

> Handle missing images (This turned out to be harder than I anticipated, probably left for future releases)

So when the text is there, and the image is not?

File: docs/multimodal.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
+# Multimodal Topic Modelling ***(BETA)***
+
+!!! note 
+    Multimodal modelling is still a BETA feature in Turftopic, and the interface is likely to change in the future.
+    We also intend on implementing multimodal modelling in a dynamic context and adding more interpretation functionality for multimodal models.
```
Comment (KennethEnevoldsen) [inline docs/multimodal.md:5]: from experience I would probably not promise features, but rather just state something like 

> Multimodal modeling is still a BETA feature in Turftopic, and it is likely that we will add more features and change the interface.

File: docs/multimodal.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+        multimodal_keynmf = KeyNMF(10, encoder="clip-ViT-B-32")
+        ```
+
+    === "MTEB/MIEB"
+        !!! tip 
+            You can find current state-of-the-art embedding models and their capabilities on the [Massive Image Embedding Benchmark leaderboard](http://mteb-leaderboard.hf.space/?benchmark_name=MIEB%28Multilingual%29).
+
+        ```bash
+        pip install "mteb<2.0.0"
```
Comment (KennethEnevoldsen) [inline docs/multimodal.md:32]: wouldn't you rather do:

```
pip install turftopic[mteb]
```

File: docs/multimodal.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        import mteb
+
+        encoder = mteb.get_model("kakaobrain/align-base")
+
+        multimodal_keynmf = KeyNMF(10, encoder="clip-ViT-B-32")
+        ```
+
+## Corpus Structure
+
+Currently all documents **have to have** an image attached to them, and only one image.
```
Comment (KennethEnevoldsen) [inline docs/multimodal.md:46]: The v2 interface of mteb will make this a lot easier.

File: mkdocs.yml [hunk 4]
```diff
@@ -11,6 +11,7 @@ nav:
     - Online Topic Modeling: online.md
     - Hierarchical Topic Modeling: hierarchical.md
     - Cross-Lingual Topic Modeling: cross_lingual.md
+    - Multimodal Modeling (BETA): multimodal.md
```
Comment (KennethEnevoldsen) [inline mkdocs.yml:14]: modelling or modeling

File: tests/test_multimodal.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+from turftopic import (
+    GMM,
+    AutoEncodingTopicModel,
+    ClusteringTopicModel,
+    KeyNMF,
+    SemanticSignalSeparation,
+)
+
+encoder = "sentence-transformers/clip-ViT-B-16"
```
Comment (KennethEnevoldsen) [inline tests/test_multimodal.py:15]: Oh, this looks like something that will make your test take quite a while to load (it will be run even if you try to run a specific test).

I would put these in a fixture.

File: tests/test_multimodal.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        feature_importance="centroid",
+        encoder=encoder,
+    ),
+]
+
+flowers = load_dataset(
+    "Multimodal-Fatima/OxfordFlowers_test", split="test"
+).train_test_split(test_size=256, seed=42)["test"]
+texts = flowers["blip_caption"]
+images = flowers["image"]
```
Comment (KennethEnevoldsen) [inline tests/test_multimodal.py:41]: These seem like quite large test objects for what you want to test. 

I would probably make a dummy datasets with 2-4 images.

File: turftopic/multimodal.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+    document_embeddings: np.ndarray
+
+
+class MultimodalModel:
+    """Base model for multimodal topic models."""
+
+    def encode_multimodal(
+        self,
+        sentences: list[str],
+        images: list[ImageRepr],
```
Comment (KennethEnevoldsen) [inline turftopic/multimodal.py:46]: if you want to allow for multiple or no images, pr. model, you probably need to make it a list of lists (but it is totally fine to change in the future).

**Review by KennethEnevoldsen — APPROVED** [review]
Looks good! Seems like most of the changes are minor and that the MultiModal encoder will likely change (but we know that).

It is probably worth noting that many want to stream images to the encoder as they take up a lot of memory, but I think that is for a future issue.

File: turftopic/multimodal.py (hunk 7 above)
Comment (x-tabdeveloping) [reply turftopic/multimodal.py:46]: I agree, I started out doing it that way, but I figured that since the MTEB interface is going to change, I will just try to follow what we do there and not put too much effort into this before the interface change.

File: mkdocs.yml (hunk 4 above)
Comment (x-tabdeveloping) [reply mkdocs.yml:14]: Depends on whether you write in American or British, unfortunately I wrote most of the docs in American, but usually write papers in British, so I confused myself here :D

File: docs/multimodal.md (hunk 3 above)
Comment (x-tabdeveloping) [reply docs/multimodal.md:46]: Yea, I've looked at the new interface proposals, I will implement features like this when that's merged.

File: docs/multimodal.md (hunk 2 above)
Comment (x-tabdeveloping) [reply docs/multimodal.md:32]: Not right now, I think it's fine for now, I will fix this once we get closer to the ideal implementation

File: tests/test_multimodal.py (hunk 5 above)
Comment (x-tabdeveloping) [reply tests/test_multimodal.py:15]: Good idea, will do!
